hal/src/stm32f2xx/gpio_hal.c

```diff
@@ -230,3 +231,37 @@ int32_t HAL_GPIO_Read(uint16_t pin)
 
     return GPIO_ReadInputDataBit(PIN_MAP[pin].gpio_peripheral, PIN_MAP[pin].gpio_pin);
 }
+
+/*
+ * @brief   blocking call to measure a high or low pulse
+ * @returns uint32_t pulse width in microseconds up to 3 seconds,
+ *          returns 0 on 3 second timeout error, or invalid pin.
+ */
+uint32_t HAL_Pulse_In(pin_t pin, uint16_t value)
+{
+    STM32_Pin_Info* SOLO_PIN_MAP = HAL_Pin_Map();
+    #define pinReadFast(_pin) ((SOLO_PIN_MAP[_pin].gpio_peripheral->IDR & SOLO_PIN_MAP[_pin].gpio_pin) == 0 ? 0 : 1)
+
+    volatile uint32_t timeoutStart = SYSTEM_TICK_COUNTER; // total 3 seconds for entire function!
+
+    /* If already on the value we want to measure, wait for the next one.
+     * Time out after 3 seconds so we don't block the background tasks
+     */
+    while (pinReadFast(pin) != value) {
+        if (SYSTEM_TICK_COUNTER - timeoutStart > 360000000UL) {
```

Comment:
I noticed these timeout values are the same on the stm32f2 and on the core - would think they would need to be different due to the different clock speed?